### ISSUE TITLE
[Feature Request] Provisioning: Allow provisioning Service Accounts (with API keys)

### ISSUE THREAD

**Issue by pdf** [grafana/grafana#82987]
Migrated from: https://github.com/grafana/grafana/discussions/49552

**Why is this needed**:

Service Accounts offer a nice method of integrating external tooling with Grafana without relying on authentication as a particular user, however the story for automated deployment is problematic.

Currently to create a service account via automation, it appears that the only method is via API calls, which requires authentication via an existing user/key, which defeats the purpose of service accounts - we're back to relying on a particular user's credentials being available for the action.

It also requires some not-insignificant plumbing - writing a client that does a multi-stage dance through the API to:

1. Check for an existing SA
2. Create the SA
3. Create an API key for the SA
4. Export the SA ID and API key for use

If I was doing this in Kubernetes, for example, I'd probably have to resort to writing a whole operator or something, backed by user credentials, to make this work with GitOps.

**What would you like to be added**:

I propose that there should be a filesystem-backed method to provision service accounts, and associated pre-defined API keys, so that user accounts and API round-trips do not need to be involved in the process.

**Who is this feature for?**

Users desiring automation of service account management.

**Comment by andaag** [comment]
Strong :heavy_plus_sign: . I'd want to create a service account token (and even rotate it) on deployment, and then provision everything else using the terraform grafana provider. Currently this requires static keys.

**Comment by Elnadrion** [comment]
+1 for the feature request

**Comment by GabrielSanchezGansinger** [comment]
This feature would really help me. 
I am trying to generate a dashboard automatically from the source of the data, and needing to get a token every time I spin up the docker is really annoying. 
Definitely a +1 to this request.

**Comment by cwrau** [comment]
This would be amazing, having to do these steps manually gets old quick 😅

**Comment by doctorjlove** [comment]
This feature would be very helpful for a fully automated provisioning solution. We use a tool to provision alerting resources via the HTTP API, and having to manually create a service account for each new instance of Grafana defeats the purpose of automation.

**Comment by Tasty213** [comment]
Getting started using grafana live and this is really key if we're going to use it at scale for IOT stuff. It's going to be a real pain having to update the api keys whenever we want to update the grafana dashboards. Was hoping to be able to have properly versioned dashboards and pass an environment variable in at startup for creating a service account for grafana live data to be sent through.

**Comment by Christian-Little-Tava** [comment]
Also interested in this

**Comment by chicagobuss** [comment]
yes plz.   I don't want to write some wonky ansible, terraform, or pulumi scripting to do something which should also be available by default on clean install with a simple config-based option.

Preferably we could just provision the entire grafana instance with a new randomly-generated admin service account token and get that as an output from terraform or whatever.

**Comment by tferrari92** [comment]
also interested!

**Comment by vpmartin** [comment]
Any news on this matter? Also in the GitOps use case, and I'll probably have to resort to a writing a full operator just for that issue.

**Comment by kalavt** [comment]
Mark

**Comment by kejrak** [comment]
Another concrete use case for this, hit operationally:

We run Grafana on Kubernetes (grafana-operator) intentionally **without persistence** — everything is reconstructed from declarative config: datasources and dashboards via provisioning/CRs, admin credentials from config. The one thing that can't be reconstructed is service account tokens. `GrafanaServiceAccount` CRs re-mint tokens into Kubernetes Secrets after every pod restart (the operator can't do better — Grafana only stores the hash), so every rollout silently invalidates every token we've handed out to consumers running outside the cluster.

Workarounds we evaluated, none of which really solves it:

- syncing the minted token Secrets outward (external-secrets `PushSecret` → vault) keeps a copy fresh, but every consumer holding the previous value still breaks on each rollout
- enabling persistence just to keep token hashes alive means keeping all UI state around, which defeats a stateless/GitOps deployment
- `auth.jwt` is the closest stateless alternative (config-driven keys + `auto_sign_up`), but then you're operating your own token issuer and lose per-token revocation
- proxy-level auth in front of Grafana hides real identities from Grafana entirely

Being able to provision a service account with a pre-defined token (or supply the secret so the hash can be seeded at startup) would make Grafana the last piece of our stack that's fully reconstructible from config.
